tests/utils/versions_test.py

```diff
@@ -15,64 +15,126 @@
+            ("7.12.1-SNAPSHOT", "with_suffix"),
+             ("7.12.1", "with_patch"),
+             ("7.12", "with_minor"),
+             ("7", "with_major")]
+        assert versions.VersionVariants("10.3.63").all_versions == [
+            ("10.3.63", "with_patch"),
+            ("10.3", "with_minor"),
+            ("10", "with_major")]
 
     def test_versions_rejects_invalid_version_strings(self):
-        with self.assertRaises(exceptions.InvalidSyntax) as ctx:
-            versions.versions("5.0.0a-SNAPSHOT")
-        self.assertEqual(r"version string '5.0.0a-SNAPSHOT' does not conform to pattern '^(\d+)\.(\d+)\.(\d+)(?:-(.+))?$'"
-                         , ctx.exception.args[0])
+        with pytest.raises(
+                exceptions.InvalidSyntax,
+                match=re.escape(r"version string '5.0.0a-SNAPSHOT' does not conform to pattern "
+                                r"'^(\d+)\.(\d+)\.(\d+)(?:-(.+))?$'")
+        ):
+            versions.VersionVariants("5.0.0a-SNAPSHOT")
 
     def test_find_best_match(self):
-        self.assertEqual("master", versions.best_match(["1.7", "2", "5.0.0-alpha1", "5", "master"], "6.0.0-alpha1"),
-                         "Assume master for versions newer than latest alternative available")
-        self.assertEqual("5", versions.best_match(["1.7", "2", "5.0.0-alpha1", "5", "master"], "5.1.0-SNAPSHOT"),
-                         "Best match for specific version")
-        self.assertEqual("master", versions.best_match(["1.7", "2", "5.0.0-alpha1", "5", "master"], None),
-                         "Assume master on unknown version")
-        self.assertIsNone(versions.best_match(["1.7", "2", "5.0.0-alpha1", "5", "master"], "0.4"), "Reject versions that are too old")
+        assert versions.best_match(["1.7", "2", "5.0.0-alpha1", "5", "master"], "6.0.0-alpha1") == "master",\
+            "Assume master for versions newer than latest alternative available"
+
+        assert versions.best_match(["1.7", "2", "5.0.0-alpha1", "5", "master"], "5.1.0-SNAPSHOT") == "5",\
+            "Best match for specific version"
+
+        assert versions.best_match(["1.7", "2", "5.0.0-alpha1", "5", "master"], None) == "master",\
+            "Assume master on unknown version"
+
+        assert versions.best_match(["1.7", "2", "5.0.0-alpha1", "5", "master"], "0.4") is None,\
+            "Reject versions that are too old"
+
+        assert versions.best_match(["7", "7.10.2", "7.11", "7.2", "5", "6", "master"], "7.10.2") == "7.10.2", \
+            "Exact match"
+
+        assert versions.best_match(["7", "7.10", "master"], "7.1.0") == "7", \
+            "Best match is major version"
+
+        assert versions.best_match(["7", "7.11", "7.2", "5", "6", "master"], "7.11.0") == "7.11",\
+            "Best match for specific minor version"
+
+        assert versions.best_match(["7", "7.11", "7.2", "5", "6", "master"], "7.12.0") == "7.11",\
+            "If no exact match, best match is the closest minor"
```

Comment:
I interpret "closest" as a synonym for "nearest" in this context. IMHO we should stick to one term consistently.

Reply:
Addressed in 861a176a7466163096ac2af17b431df48d7e2a3f